src/Npgsql/NpgsqlBatch.cs

```diff
@@ -171,4 +171,30 @@ public override Task PrepareAsync(CancellationToken cancellationToken = default)
 
     /// <inheritdoc />
     public override void Cancel() => Command.Cancel();
+
+    /// <inheritdoc />
+    public override void Dispose()
+    {
+        ClearTransaction();
+        if (Command.IsCached && Connection is not null && Connection.CachedBatch is null)
+        {
+            BatchCommands.Clear();
+            Connection.CachedBatch = this;
+            return;
+        }
+
+        Command.IsCached = false;
+    }
+
+    /// <summary>
+    /// Clears the current transaction.
+    /// </summary>
+    protected virtual void ClearTransaction() => Transaction = null;
```

Comment:
It's unfortunate this has to be a bit convoluted like this (because of NpgsqlDataSourceBatch)... But I can't really think of a better way.

Reply:
Maybe expose ResetTransaction on NpgsqlCommand, which just sets _transaction to null. Then we can just call that directly from NpgsqlBatch.Dispose?

Reply:
You don't like this suggestion?